docs/design/interceptors.md

```diff
@@ -0,0 +1,691 @@
+The `Interceptor` interface defined above is generic over:
+* `Input` - the operation input type
+* `Output` - the operation output type
+* `ProtocolRequest` - the specific protocol (transmit) request type
+* `ProtocolResponse` - the specific protocol (transmit) response type
+
+
+Observations:
+
+* Every interceptor method has a default implementation. This allows customers to override only the hooks they are 
+interested in as well as allowing future hooks to be defined without breaking API or binary compatibility.
+* Multiple context types are defined that restrict the set of information available to a particular hook (e.g. you can't
+access the operation'output in any hook until *after* the first deserialization hook). Additionally the hooks are 
+specific about the nullability of particular context information (e.g. the `modifyBeforeCompletion` hook may or may not
+have a protocol request or response available depending on how far the operation execution made it.
+* Hooks are non-blocking (note no `suspend` in any hook method) and implementators are expected to respect this
+* Despite being generic over the input and output type, the `Input` and `Output` types of an `Interceptor` are expected
+to be `Any` in practice. This is because interceptors can be registered at the *client* level. Such an interceptor would
+be executed for every operation and would not be able to know the specific input or output type at compile time. 
+Interceptors that execute for any operation are common and expected (e.g. add a header to every outgoing request). 
+The downside to this flexibility is that customers will have to cast the input or output type from the context to the 
+specific operation type. 
+  * NOTE: The modeled operation name will be available in the context (it is already available in the execution context)
+  which can be used on interceptors registered at the client level that want to deal with specific operations.
+
+
+### HTTP Interceptor
+
+
+The `Interceptor` interface defined above would not be useful to a customer without specific types. Each protocol will
+get it's own `typealias` that fills in the generics that make sense for a particular protocol. For instance HTTP based 
+protocol clients would be generated to reference the more specific interceptor type:
+
+```kotlin
+typealias HttpInterceptor = Interceptor<Any, Any, HttpRequest, HttpResponse>
+```
+
+
+#### Alternative 1
+
+Alternatively the `Interceptor` interface could hard code `Any` for `Input` and `Output` types. This would simplify 
+some of the runtime type definitions by removing the propagation of those generics. The end result to the customer would
+be the same (a protocol specific interceptor `typealias` is still expected). 
+
+
+## Interceptor Registration
+
+Interceptors will be able to be registered at client creation time as well as through runtime plugins (TBD).
+
+
+Generated client config will be modified to include a list of interceptors that can be appended to:
```

Comment:
**Comment**: May want to include a note indicating that these config-specified interceptors are _strictly additive_ to any pre-existing, non-configurable interceptors.